File: project_euler/Problem023/Problem023.py
import math

class Problem023:
    def solution(self) -> None:
        start, end, abundants = 24, 28123, set()
        for i in range(start >> 1, start + 1):
            if self.__is_abundant(i):
                abundants.add(i)

        result = start * (start - 1) >> 1
        for i in range(start + 1, end + 1):
            is_non_abundant = True
            for abundant in abundants:
                if i - abundant in abundants:
                    is_non_abundant = False
                    break
            if is_non_abundant:
                result += i

            if self.__is_abundant(i):
                abundants.add(i)
        print(result)

    def __is_abundant(self, value: int) -> bool:
        border, summation = int(math.sqrt(value)), 1
        for i in range(2, border):
            if not value % i:
                summation += i + value // i
        if not value % border:
            summation += border
            if value != border * border:
                summation += value // border
        return summation > value

File: project_euler/Problem023/Problem023.go
package project_euler

import (
    "fmt"
    "math"
)

func Problem023() {
    start := 24
    end := 28123
    abundants := map[int]bool{}
    for i := start >> 1; i <= start; i++ {
        if isAbundant(i) {
            abundants[i] = true
        }
    }

    result := start * (start - 1) >> 1
    for i := start + 1; i <= end; i++ {
        isNonAbundant := true
        for abundant := range abundants {
            if abundants[i - abundant] {
                isNonAbundant = false
                break
            }
        }
        if isNonAbundant {
            result += i
        }

        if isAbundant(i) {
            abundants[i] = true
        }
    }
    fmt.Println(result)
}

func isAbundant(value int) bool {
    border := int(math.Sqrt(float64(value)))
    summation := 1
    for i := 2; i < border; i++ {
        if 0 == value % i {
            summation += i + value / i
        }
    }
    if 0 == value % border {
        summation += border
        if value != border * border {
            summation += value / border
        }
    }
    return summation > value
}
